### PR TITLE
Add dynamic run priorities

### PR DESCRIPTION
**Pull request by tonyhb**
This PR introduces run priority factors to individual function runs. This allows people to specify priorities to runs within the same function. For example, within the same function users may want to prioritize runs from paid users.

A non-zero priority.run expression result changes when a function run should execute.  A value > 0 has increased priority. A value < 0 has lower priority and will delay functions from running for the specified time in seconds.

For example:  if the expression returns 120, the function run will run before any other work enqueued in the last 120 seconds.  At scale, with a high backlog, this ensures that work > 120 seconds still runs and that this new job runs before any work enqueued within the last 120 seconds.

Note: this does not apply to jobs scheduled in the future.

## Type of change (choose one)
- [ ] Chore (refactors, upgrades, etc.)
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Security fix (non-breaking change that fixes a potential vulnerability)
- [ ] Docs
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)

## Checklist
- [ ] I've linked any associated issues to this PR.
- [ ] I've tested my own changes.

*[Check our Pull Request Guidelines](https://github.com/inngest/inngest/blob/main/docs/PULL_REQUEST_GUIDELINES.md)*

### CODE REVIEW

File: pkg/execution/state/state.go [hunk 1]
```diff
@@ -65,6 +65,9 @@ type Identifier struct {
 	WorkspaceID uuid.UUID `json:"wsID"`
 	// If this is a rerun, the original run ID is stored here.
 	OriginalRunID *ulid.ULID `json:"oRunID,omitempty"`
+	// PriorityFactor is the overall priority factor for this particular function
+	// run.  This allows individual runs to take precedence within the same queue.
+	PriorityFactor *int64 `json:"pf,omitempty"`
```
Comment (darwin67) [inline pkg/execution/state/state.go:70]: Better to provide a note what the `int64` is for.

e.g. The higher the number, the higher the priority within the range.

Probably don't need an essay here describing it as the seconds, but good to have a little note here so for people reading the code, we know this number actually represent seconds.

**Review by darwin67 — APPROVED** [review]
Looking good. Mostly minor comments.

File: tests/golang/priority_test.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+		require.NoError(t, err)
+		// The function takes 5s to run, so wait a second ensuring that we run events in
+		// order and start processing functions.  By the first event, we have a backlog.
+		<-time.After(2 * time.Second)
+	}
+
+	require.Eventually(t, func() bool {
+		return atomic.LoadInt32(&counter) == 3
+	}, 20*time.Second, time.Second)
+}
```
[the remark this answers is not in the document]
Comment (darwin67) [reply tests/golang/priority_test.go:92]: Ah I just saw you added more tests, so this comment probably can be ignored.

File: pkg/execution/executor/executor.go [hunk 3]
```diff
@@ -250,7 +250,15 @@ func (e *executor) Schedule(ctx context.Context, req execution.ScheduleRequest)
 		mapped[n] = item.GetEvent().Map()
 	}
 
-	_, err := e.sm.New(ctx, state.Input{
+	factor, err := req.Function.RunPriorityFactor(ctx, mapped[0])
+	if err != nil && e.log != nil {
+		e.log.Warn().Err(err).Msg("run priority errored")
```
[the remark this answers is not in the document]
Comment (tonyhb) [reply pkg/execution/executor/executor.go:255]: 100% - was thinking the same thing.  let's hook in an inngest event and then use the platform to send our users alerts?

[the remark this answers is not in the document]
Comment (darwin67) [reply pkg/execution/executor/executor.go:255]: That, and also how to surface it in the UI.
Maybe the event need to have the runID, and some additional metadata so we know how to connect that info.

File: pkg/inngest/function.go [hunk 4]
```diff
@@ -232,6 +250,44 @@ func (f Function) Validate(ctx context.Context) error {
...
+	if f.Priority == nil || f.Priority.Run == nil {
+		return 0, nil
+	}
+
+	expr, err := expressions.NewExpressionEvaluator(ctx, *f.Priority.Run)
+	if err != nil {
+		return 0, fmt.Errorf("Priority.Run expression is invalid: %s", err)
+	}
+
+	val, _, err := expr.Evaluate(ctx, expressions.NewData(map[string]any{"event": event}))
```
[the remark this answers is not in the document]
Comment (tonyhb) [reply pkg/inngest/function.go:264]: Luckily these expressions are fully cacheable as they're not dynamic!  When you deploy a fn the expression stays the same for the entire lifetime, so it's a constant time operation :D

File: tests/golang/priority_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tonyhb) [reply tests/golang/priority_test.go:92]: Yeah sorry, had that break in between!

File: pkg/execution/executor/executor.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (tonyhb) [reply pkg/execution/executor/executor.go:255]: When we do tagging we'll add system level warns;  that needs to be specced with John.  It's a good candidate for all things "bad but not critical".

**Pull request merged**